gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/UserResource.java

```diff
@@ -174,6 +168,35 @@ public void updateUserStatus(
                 .subscribe(response::resume, response::resume);
     }
 
+    @PATCH
+    @Path("/username")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            nickname = "updateUsername",
+            value = "Update a user username",
+            notes = "User must have the DOMAIN_USER[UPDATE] permission on the specified domain " +
+                    "or DOMAIN_USER[UPDATE] permission on the specified environment " +
+                    "or DOMAIN_USER[UPDATE] permission on the specified organization")
+    @ApiResponses({
+            @ApiResponse(code = 201, message = "User username successfully updated", response = User.class),
+            @ApiResponse(code = 500, message = "Internal server error")})
+    public void updateUsername(
+            @PathParam("organizationId") String organizationId,
+            @PathParam("environmentId") String environmentId,
+            @PathParam("domain") String domain,
+            @PathParam("user") String userId,
+            @ApiParam(name = "username", required = true) @Valid @NotNull UsernameEntity username,
+            @Suspended final AsyncResponse response) {
+        final io.gravitee.am.identityprovider.api.User authenticatedUser = getAuthenticatedUser();
+
+        checkAnyPermission(organizationId, environmentId, domain, Permission.DOMAIN_USER, Acl.UPDATE)
+                .andThen(domainService.findById(domain)
+                        .switchIfEmpty(Maybe.error(new DomainNotFoundException(domain)))
+                        .flatMapSingle(irrelevant -> userService.updateUsername(ReferenceType.DOMAIN, domain, userId, username.getUsername(), authenticatedUser)))
+                .subscribe(response::resume, response::resume);
+    }
```

Comment:
Should we return maybe only the updated username instead of the entire user ?

Reply:
We could, but `updateStatus` returns the whole user. I don't feel strongly either way.

Reply:
Agreed, let's roll with John